ultravox/tools/infer_tool.py

```diff
@@ -190,25 +192,44 @@ def dataset_infer(inference: base.VoiceInference, args: InferArgs):
     if args.seed is not None:
         ds_args.shuffle_seed = args.seed
     ds = datasets.create_dataset(args.data_sets[0], ds_args)
-    scores: List[float] = []
-    for i, sample in enumerate(datasets.Range(ds, args.num_samples)):
-        # Store the original question and answer for JSON output.
-        question_text = sample.audio_transcript
-        expected_answer = sample.messages[-1]["content"]
-        # Drop any assistant response from the sample.
-        sample.messages = sample.messages[:-1]
-        if not args.json:
-            run_tui(i, inference, sample, args, expected_answer, scores)
-        else:
-            output = inference.infer(
-                sample, max_tokens=args.max_tokens, temperature=args.temperature
-            )
-            obj = {
-                "question": question_text,
-                "generated_answer": output.text,
-                "expected_answer": expected_answer,
-            }
-            print(json.dumps(obj))
+
+    if args.batch_size:
+        start_time = time.time()
+        samples = []
+        for i, sample in enumerate(datasets.Range(ds, args.num_samples)):
+            samples.append(sample)
+        samples = sorted(samples, key=lambda sample: len(sample.audio_transcript))
+        output = inference.batch_infer(
+            samples,
+            batch_size=args.batch_size,
+            max_tokens=args.max_tokens,
+            temperature=args.temperature,
+        )
+        print("printing outputs...")
+        for sample in output:
+            print(sample)
+        print("Total time", time.time() - start_time)
+
+    else:
+        scores: List[float] = []
+        for i, sample in enumerate(datasets.Range(ds, args.num_samples)):
+            # Store the original question and answer for JSON output.
+            question_text = sample.audio_transcript
+            expected_answer = sample.messages[-1]["content"]
+            # Drop any assistant response from the sample.
+            sample.messages = sample.messages[:-1]
+            if not args.json:
+                run_tui(i, inference, sample, args, expected_answer, scores)
+            else:
+                output = inference.infer(
+                    sample, max_tokens=args.max_tokens, temperature=args.temperature
+                )
```

Comment:
I suggest we combine `batch` mode and `json` mode together if you can. Now there are 3 paths for the code to take which seems like too many.

Reply:
Using [BatchSampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.BatchSampler) might help.